File: atcoder/C_Large_Queue.cpp
#include <bits/stdc++.h>

using namespace std;
using ll = unsigned long long;

const int MAXN = 3e5 + 2;
ll c1[MAXN], c2[MAXN], c3[MAXN];

int main(){
    int n; cin >> n;
    int len = 1, rem = 0;
    ll remsum = 0, ans;
    ll t, c, x, k;

    while(n--){
        cin >> t;
        if (t == 1){
            cin >> c >> x;
            c1[len] = x;
            c2[len] = c + c2[len - 1];
            c3[len] = x * c + c3[len - 1];
            len++;
        }else{
            cin >> k;
            rem += k;
            size_t idx = upper_bound(c2, c2 + len, rem) - c2;
            ans = c3[idx - 1] + c1[idx] * (rem - c2[idx-1]) - remsum;
            cout << ans << '\n';
            remsum += ans;
        }
    }
    return 0;
}

File: atcoder/C_Large_Queue.py
from bisect import bisect_left

n = int(input())
c1,c2,c3 = [0], [0],[0]
remsum = 0
total = 0

for _ in range(n):
    t = [*map(int,input().split())]

    if t[0] == 1:
        c1.append(t[2])
        c2.append(t[1] + c2[-1])
        c3.append(t[1] * t[2] + c3[-1])
    else:
        total += t[1]
        idx = bisect_left(c2,total)
        ans = c1[idx] * (total - c2[idx-1]) - remsum + c3[idx-1]
        remsum += ans
        print(ans)